add new vep benchmark

naliboff opened the pull request:
This PR adds a new benchmark from Gerya 2019 (Introduction to Numerical Geodynamic Modeling), which examines shortening of a visco plastic or viscoelastic plastic block under compression. This benchmarking will provide a unique point of comparison for #2532 and #2314, and joint work with @alarshi, @cedrict, @jdannberg, @bobmyhill, and others. 

A number of questions and decisions before proceeding further:

1. Should we provide a unique folder for each set of benchmarks in this folder? I think the answer is likely yes if we add any additional files?
2. Should we a README.md file to the main folder and any sub-folders? There is already quite a bit of documentation in the prm files, but this would be useful if we add additional files to sub-folders.
3. The points value post processor unfortunately does not output strain rate. Should I add a python script that reads in VTU data and then plots a strain rate profile?

* [x] I have followed the [instructions for indenting my code](../blob/master/CONTRIBUTING.md#making-aspect-better).
* [X] I have tested my new feature locally to ensure it is correct.
* [x] I have [created a testcase](http://www.math.clemson.edu/~heister/manual.pdf#sec%3Awriting_tests) for the new feature/benchmark in the [tests/](../blob/master/tests/) directory.
* [x] I have added a changelog entry in the [doc/modules/changes](../blob/master/doc/modules/changes) directory that will inform other users of my change.

bangerth commented:
1. Yes.
2. Yes, a short README.md always makes sense.
3. That would actually be quite a neat idea -- it would be a nice opportunity to introduce another way of postprocessing information.

bangerth commented on benchmarks/viscoelastic_plastic_shear_bands/gerya_2019_vp.prm:
```diff
@@ -0,0 +1,196 @@
+  end
+  subsection Newton solver parameters
+    set Max Newton line search iterations        = 0
+    set Max pre-Newton nonlinear iterations      = 500 # Ensures only defect-correct Picard method is used
```
```suggestion
    set Max pre-Newton nonlinear iterations      = 500 # Ensures only defect-correction Picard method is used
```

bangerth commented on benchmarks/viscoelastic_plastic_shear_bands/gerya_2019_vp.prm:
```diff
@@ -0,0 +1,196 @@
+  end
+end
+
+# Globally refine to the mesh. A minimum of 3 global refinements
```
```suggestion
# Globally refine the mesh. A minimum of 3 global refinements
```

bangerth commented on benchmarks/viscoelastic_plastic_shear_bands/gerya_2019_vp.prm:
```diff
@@ -0,0 +1,196 @@
+  subsection Function
+    set Variable names      = x,y
+    set Function constants  = vel=0.15778463
+    set Function expression = if (x<50e3 , vel, -vel); if (y<50e3, -vel, vel);
```
```suggestion
    set Function expression = if (x<50e3, vel, -vel); if (y<50e3, -vel, vel);
```

tjhei commented on benchmarks/viscoelastic_plastic_shear_bands/README.md:
```diff
@@ -0,0 +1,14 @@
+The folders in this directory contain multiple benchmarks for examing the 
```
```suggestion
The folders in this directory contain multiple benchmarks for examining the 
```

tjhei commented on benchmarks/viscoelastic_plastic_shear_bands/gerya_2019/README.md:
```diff
@@ -0,0 +1,22 @@
+
+- gerya_2019_vep.prm: Parameter file that is derived from 
+gera_2019_vep.prm and modifies the rheological model to
+incldue viscoelasticity (i.e., viscoelastic-plastic rheology).
```
```suggestion
include viscoelasticity (i.e., viscoelastic-plastic rheology).
```

bobmyhill commented on benchmarks/viscoelastic_plastic_shear_bands/gerya_2019/gerya_2019_vp.prm:
```diff
@@ -0,0 +1,196 @@
+# provided by gerya_2019_vep.prm within this folder.
+#
+# In detail, the benchmark examines shortening of a viscoelastic
+# plastic (here, visco plastic) block, which contains a weak 
```
```suggestion
# plastic (here, visco-plastic) block, which contains a weak 
```
(for consistency)

bobmyhill commented on benchmarks/viscoelastic_plastic_shear_bands/gerya_2019/gerya_2019_vep.prm:
```diff
@@ -0,0 +1,74 @@
+    set Maximum viscosity           = 1.e23
+    set Minimum viscosity           = 1.e17
+
+    # Nonsensical Prefactor values (1e-50) are assigned to the compositional fields tracking elastic stresses
```
explain why?

naliboff commented:
@bangerth @tjhei @bobmyhill - Thanks for the preliminary reviews! All of the necessary files, including a python script, are now present and this should be ready for a second review.

tjhei commented on benchmarks/viscoelastic_plastic_shear_bands/gerya_2019/gerya_2019_analysis.py:
```diff
@@ -0,0 +1,100 @@
+# Python script to load the solution data into a
+# numpy array, interpolate the data to a new
+# uniform grid, and plot the results.
+# Within a python interpreter, the script can 
```
This is in ParaView or a standalone python code? Why don't you say ``python gerya_2019_analysis.py``?

naliboff replied:
That would work as well. The command I listed is for when you want to execute the script while working inside ipython (or similar). I find this helpful for doing quick debugging and testing without needing to rerun the whole script. In any case, good to spell out both options, which is now included in the documentation. In the last commit I also specified the versions of each package in the documentation.

tjhei commented on benchmarks/viscoelastic_plastic_shear_bands/gerya_2019/gerya_2019_analysis.py:
```diff
@@ -0,0 +1,109 @@
+# numpy array, interpolate the data to a new
+# uniform grid, and plot the results.
+# From a terminal, execute the script with
+#   python gerya_2019_analysis.py. 
```
```suggestion
#   python gerya_2019_analysis.py
```

tjhei commented on benchmarks/viscoelastic_plastic_shear_bands/gerya_2019/gerya_2019_analysis.py:
```diff
@@ -0,0 +1,109 @@
+# uniform grid, and plot the results.
+# From a terminal, execute the script with
+#   python gerya_2019_analysis.py. 
+# Within a python interpreter (e.g., ipython), 
```
```suggestion
# Within a python interpreter (e.g., a Jupyter notebook), 
``j

tjhei replied:
ipython? You are living in the past! :-)

naliboff commented:
@tjhei - you have no idea, it is VI on side of the screen and ipython on the other :-)

I've addressed the remaining comments, but did notice that two of the testers are failing. I thinking this is due to moving files around associated with the tests and cookbooks, but it is not occurring for the no-unity build case. Aside from resolving this, any additional requests for changes?

tjhei commented:
> I thinking this is due to moving files around associated with the tests and cookbooks, but it is not occurring for the no-unity build case. Aside from resolving this, any additional requests for changes?

The error is
```
File '*.prm' missing in /jenkins/workspace/aspect_PR-3825/benchmarks/viscoelastic_plastic_shear_bands.
```
Can you adjust https://github.com/geodynamics/aspect/blob/master/benchmarks/check.mk by adding a section for viscoelastic_plastic_shear_bands? Let me know if you need helpl,.

tjhei commented:
Something like:
``
viscoelastic_plastic_shear_bands/: dummy
	@$(def); run_all_prms $@/gerya_2019
	@$(def); run_all_prms $@/anotherfolder
```

tjhei commented:
did you figure out what to do, @naliboff ?

naliboff commented:
@tjhei - Sorry for leaving this hanging for so long and hopefully I've added the right changes to check.mk. I also added two commits that enable using material averaging with viscoelasticity (viscosity averaging only options) and an associated test.

tjhei commented:
this looks good, but you need to update the test results (screen output changed on master)

tjhei commented:
I am happy. @jdannberg ?

jdannberg commented on benchmarks/viscoelastic_plastic_shear_bands/gerya_2019/gerya_2019_analysis.py:
```diff
@@ -0,0 +1,109 @@
+yi = np.linspace(np.min(y), np.max(y), int((np.max(y) - np.min(y))/res)*2 + 1 )
+
+# Generate data interpolation grid. This is a helpful step as the VTU data format
+# has more than value specified at each grid point. This is due to setting
```
more than one value?

naliboff replied:
Yes, when I looked at the raw data you can see the same coordinate listed more than once and having a different value. This may be due to the reason you specified below. In any case, I remove the associated text to avoid confusion.

jdannberg commented on benchmarks/viscoelastic_plastic_shear_bands/gerya_2019/gerya_2019_analysis.py:
```diff
@@ -0,0 +1,109 @@
+
+# Generate data interpolation grid. This is a helpful step as the VTU data format
+# has more than value specified at each grid point. This is due to setting
+# 'set Interpolate output = true' in ASPECT.
```
I thought that was due to the general format of how we write the output (we write out unstructured meshes, and so we have to give the locations of the corners of each cell). Bt maybe you don't mean what I think you mean here.

naliboff replied:
You are probably correct that vertices are repeated even without the interpolate output option. Regardless, in the raw VTU data specific points (vertices) are list more than once, and as you stated this is almost certainly due to the way we write out unstructured data. Interpolating the output in ASPECT should then just cause further repetition of some vertices. In any case, as noted above, I removed this comment to avoid confusion :)

jdannberg commented on benchmarks/viscoelastic_plastic_shear_bands/gerya_2019/gerya_2019_vep.prm:
```diff
@@ -0,0 +1,75 @@
+include $ASPECT_SOURCE_DIR/benchmarks/viscoelastic_plastic_shear_bands/gerya_2019/gerya_2019_vp.prm
+
+#  Global parameters
+set Dimension                              = 2
```
You don't need this, correct? (it's already in the other file you include)

naliboff replied:
yep, removed.

jdannberg commented on benchmarks/viscoelastic_plastic_shear_bands/gerya_2019/gerya_2019_vep.prm:
```diff
@@ -0,0 +1,75 @@
+    set Reference viscosity         = 1.e21
+    set Reference strain rate       = 1.e-13
+    set Maximum viscosity           = 1.e23
+    set Minimum viscosity           = 1.e17
```
The five parameters above are the same as in the original file, so you don't need them here. 
(unless you think it's important to list them as a reminder for the user)

naliboff replied:
Good catch, no reason to keep these in.

jdannberg commented on benchmarks/viscoelastic_plastic_shear_bands/gerya_2019/gerya_2019_vp.prm:
```diff
@@ -0,0 +1,196 @@
+  set Temperature polynomial degree     = 1
+end
+
+# Velocity boundary conditions (5e-9 m/s on each boundary)
```
This is just a thought for next time (you don't actually have to change anything). But if you want to prescribe the velocity in m/s, why didn't you just set Use years in output instead of seconds to false? The strain rate output won't change.

naliboff replied:
Yep, removed!

jdannberg commented on benchmarks/viscoelastic_plastic_shear_bands/gerya_2019/gerya_2019_vp.prm:
```diff
@@ -0,0 +1,196 @@
+  end
+end
+
+# Composition boundary conditions
```
I don't think you need this whole section if you don't actually have any boundary conditions for the composition.

naliboff replied:
Yep, this could be deleted. However, I like to keep this section in as a reminder that the compositional fields in fact have no boundary conditions. Purely there to avoid confusion (on my end) later on :)

jdannberg reviewed (commented):
This looks good to me! I've made a few suggestions, but I don't want to hold up the merging of this any longer, so feel free to ignore them (they are only minor points) or merge after you've fixed them. 
Sorry this took me so long, and @naliboff thank you for contributing this!

naliboff commented:
@jdannberg - Thanks for the review and catches! I addressed the comments in the last commit, and hopefully nothing changes in the tester. Once this is merged, I'll rebase the plastic damper PR and get that ready for review.

naliboff commented:
@jdannberg - Ok to merge with the last commit?
